### PR TITLE
INFRA-237: Use Qemu to virtualize RPI machine

### PR DESCRIPTION
**Pull request by chinzou**
https://mekomsolutions.atlassian.net/browse/INFRA-237

### CODE REVIEW

File: roles/bootstrap/tasks/main.yml [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+  package:
+    name: kpartx
+    state: present
+
+- name: Install pip package
+  package:
+    name: python3-pip
+    state: present
+
+- name: Check that the somefile.conf exists
```
Comment (rbuisson) [inline roles/bootstrap/tasks/main.yml]: replace by "os-origin.img"

File: roles/bootstrap/tasks/main.yml [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+  stat:
+    path: /opt/os-origin.img
+  register: unziped_image
+
+- name: Unarchive the image
+  shell: "unxz --keep /opt/os-origin.img.xz"
+  when: unziped_image.stat.exists == false
+
+- name: Install psutil python package
```
Comment (rbuisson) [inline roles/bootstrap/tasks/main.yml]: Install 'psutil' Python package

File: roles/bootstrap/tasks/main.yml [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+
+- name: Install Kpartx package
+  package:
+    name: kpartx
+    state: present
+
+- name: Install pip package
+  package:
+    name: python3-pip
+    state: present
```
Comment (rbuisson) [inline roles/bootstrap/tasks/main.yml]: We should add: `    update_cache: yes`

Comment (rbuisson) [inline roles/bootstrap/tasks/main.yml]: Let's group those 2 in 1 block:
```
- name: Install system packages
  package:
    name:
    - kpartx
    - python3-pip
    update_cache: yes
```

File: roles/bootstrap/tasks/main.yml [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+
+- name: Unarchive the image
+  shell: "unxz --keep /opt/os-origin.img.xz"
+  when: unziped_image.stat.exists == false
+
+- name: Install psutil python package
+  pip:
+    name: psutil
+
+- name: Check if VM ssh port is open
```
Comment (rbuisson) [inline roles/bootstrap/tasks/main.yml:31]: The title is a bit misleading as we are here talking about the Qemu VM, but at this stage there is nothing related to Qemu so the user don't really understand what it is.

File: roles/bootstrap/tasks/main.yml [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+    dest: /opt/initrd.img
+    remote_src: yes
+
+- name: Copy vmlinux
+  copy:
+    src: "{{ image_mount_point }}/boot/vmlinuz"
+    dest: /opt/vmlinuz
+    remote_src: yes
+
+- name: Create a user for the cluster
```
Comment (rbuisson) [inline roles/bootstrap/tasks/main.yml]: "Create the '{{ cluster_user }}' user"

File: roles/bootstrap/tasks/main.yml [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+  user:
+    name: "{{ cluster_user }}"
+    comment: Cluster User with root access
+    uid: 1041
+    group: root
+
+- name: Create cluster user in chroot filesystem
```
Comment (rbuisson) [inline roles/bootstrap/tasks/main.yml]: "Create the '{{ cluster_user }}' user in the chroot filesystem"

File: roles/bootstrap/tasks/main.yml [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+  shell: chroot {{ image_mount_point }} useradd -u 1041 -m {{ cluster_user }}
+  become: yes
+
+- name: "Create home folder"
```
Comment (rbuisson) [inline roles/bootstrap/tasks/main.yml]: "Create home folder for {{ cluster_user }}"

File: roles/bootstrap/tasks/main.yml [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+  file:
+    path: "{{ image_mount_point }}/home/{{ cluster_user }}"
+    state: directory
+    owner: "{{ cluster_user }}"
+
+- name: "Create '.ssh/' folder"
```
Comment (rbuisson) [inline roles/bootstrap/tasks/main.yml]: "Create '.ssh/' folder for '{{ cluster_user }}' user"

File: roles/bootstrap/tasks/main.yml [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+  file:
+    path: "{{ image_mount_point }}/home/{{ cluster_user }}/.ssh"
+    state: directory
+    owner: "{{ cluster_user }}"
+
+- name: Add user to sudoers
```
Comment (rbuisson) [inline roles/bootstrap/tasks/main.yml]: "Add '{{ cluster_user }}' user to sudoers"

File: roles/bootstrap/tasks/main.yml [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+  lineinfile:
+    path: "{{ image_mount_point }}/etc/sudoers"
+    line: "{{ item }}"
+  with_items:
+    - "{{ cluster_user }} ALL=(ALL) NOPASSWD: ALL"
+ 
+- name: Add ssh key to cluster user
```
Comment (rbuisson) [inline roles/bootstrap/tasks/main.yml]: "Add SSH key to '{{ cluster_user }}' user"

File: roles/common/tasks/main.yml [hunk 11]
```diff
@@ -1,11 +1,13 @@
...
+- name: Install network tools
   apt:
-    update-cache: yes
+    name: net-tools
+    state: present
 
-- name: Set the hostname as inventory name
-  hostname:
-    name: "{{inventory_hostname}}"
+- name: Add hostname to hosts
```
Comment (rbuisson) [inline roles/common/tasks/main.yml]: Could you remind me why is this needed?
Is the value of `ansible_hostname` of any importance?

File: roles/k3s/install/tasks/main.yml [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+    value: "1"
+    state: present
+    reload: yes
+
+- name: Delete k3s if already present
+  file:
+    path: /usr/local/bin/k3s
+    state: absent
+
+- name: Download k3s binary arm64
```
Comment (rbuisson) [inline roles/k3s/install/tasks/main.yml]: Download K3s binary for 'arm64' architecture

File: roles/k3s/install/tasks/main.yml [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+    reload: yes
+
+- name: Enable IPv6 forwarding
+  sysctl:
+    name: net.ipv6.conf.all.forwarding
+    value: "1"
+    state: present
+    reload: yes
+
+- name: Delete k3s if already present
```
Comment (rbuisson) [inline roles/k3s/install/tasks/main.yml]: exact naming is K3s.
If you are reffering to a package name, then let's put it in single quotes.

"Delete K3s if already present"
"Delete 'k3s' if already present"

File: roles/qemu/tasks/main.yml [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+- name: Unmount the image
+  ansible.posix.mount:
+    path: /tmp/os
+    state: unmounted
+
+- name: Run Qemu VM
+  shell: qemu-system-aarch64 -M virt -m 1024 -smp 4 -cpu cortex-a53 -kernel /opt/vmlinuz -initrd /opt/initrd.img -net tap -net nic -drive file=/opt/os.img,if=none,id=drive0,cache=writeback -device virtio-blk-device,drive=drive0,bootindex=0 -append 'root=/dev/vda2 noresume rw cgroup_enable=cpuset cgroup_memory=1 cgroup_enable=memory' -no-reboot -netdev user,id=net1,hostfwd=tcp::{{ image_ssh_port }}-:22 -device virtio-net-device,netdev=net1 -daemonize -display none
+  become: true
+
+- name: Wait for ssh port to become open on the host
```
Comment (rbuisson) [inline roles/qemu/tasks/main.yml]: Could you clarify this task name?

File: roles/common/tasks/main.yml (hunk 11 above)
Comment (rbuisson) [reply roles/common/tasks/main.yml]: Actually I think this is misplaced.
It seems that we are running the commons on k3s_cluster hosts.
However the path "{{ image_mount_point }}/etc/hosts" don't exist within the OS. Shouldn't it be run from outside the OS, directly on the host VM instead?

File: README.md [hunk 15]
```diff
@@ -1,11 +1,18 @@
 ### appliance-os
 
-Ansible playbooks to configure a Kubernetes cluster.
+Ansible playbooks to configure a Kubernetes cluster node/master.
 
-Edit the [hosts.yml](hosts.yml) file with the correct IPs and users.
 
-Run the playbook:
```
Comment (rbuisson) [inline README.md:7]: Could you add the following:

Start Multipass machine locally:
```
multipass launch -d 16G -m 4G -n os-builder -c 4
```
Paste your public key in:
```
multipass exec os-builder nano .ssh/authorized_keys
```
Retrieve the machine IP and set it in [hosts.ini](inventory/sample/hosts.ini)
```
multipass info os-builder | grep IPv4
```

**Review by rbuisson — CHANGES_REQUESTED** [review]
Thanks Zoheir!
Few remarks and changes.

File: roles/common/tasks/main.yml (hunk 11 above)
Comment (chinzou) [reply roles/common/tasks/main.yml]: This task is for K3s because it uses the hostname to check the node is up, but yes the path should be fixed.

File: roles/k3s/install/tasks/main.yml [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+  sysctl:
+    name: net.ipv6.conf.all.forwarding
+    value: "1"
+    state: present
+    reload: yes
+
+- name: "Delete 'k3s' if already present"
+  file:
+    path: /usr/local/bin/k3s
+    state: absent
```
Comment (rbuisson) [inline roles/k3s/install/tasks/main.yml]: Not sure why this is present. The next task should just redownload the file if it's missing or not the same isn't it?
Actually you would need to provide a checksum of that file to avoid re-downloading it.

File: roles/k3s/install/tasks/main.yml [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+
+- name: Download K3s binary for 'arm64' architecture
+  get_url:
```
Comment (rbuisson) [inline roles/k3s/install/tasks/main.yml:18]: so here, please provide the checksum as:
```    checksum: "{{ k3s_binary_checksum }}"```
(var added in all.yml)

File: roles/k3s/master/tasks/main.yml [hunk 18]
```diff
@@ -0,0 +1,17 @@
...
+    dest: "{{ systemd_dir }}/k3s.service"
+    owner: root
+    group: root
+    mode: 0755
+
+- name: Enable and check K3s service
+  systemd:
+    name: k3s
+    daemon_reload: yes
+    state: restarted
```
Comment (rbuisson) [inline roles/k3s/master/tasks/main.yml]: why is this state as "restarted"? Should probably be simply "started" isn't it?

File: roles/k3s/master/tasks/main.yml [hunk 19]
```diff
@@ -0,0 +1,17 @@
...
+- name: Copy K3s service file
+  register: k3s_service
+  template:
+    src: "k3s.service.j2"
+    dest: "{{ systemd_dir }}/k3s.service"
+    owner: root
+    group: root
+    mode: 0755
+
+- name: Enable and check K3s service
```
Comment (rbuisson) [inline roles/k3s/master/tasks/main.yml]: Let's rename to: "Ensure K3s service is enabled and started"
Remember Ansible is desired-state. All tasks should as much as possible be there to ensure that something is in the state desired, not to systematically change the state of things. That's why it's good to provide checksum to get_url, avoid using actions (restart, reload, start...) and rather use states (started, enabled...)

However, naming all tasks with "Ensure blabla..." ends up being heavy to the reader, but in your mind you should see it this way. For instance:
"Download K3s binary for 'arm64' architecture" is "Ensure K3s binary for 'arm64' architecture is downloaded"

File: inventory/sample/hosts.ini [hunk 20]
```diff
@@ -0,0 +1,11 @@
+[host_vm]
+rpi-creation    ansible_host=10.182.31.179   ansible_port=22     ansible_user=ubuntu
```
Comment (rbuisson) [inline inventory/sample/hosts.ini]: Let's rename this to `vm` and make sure spaces between arguments are consistent.
I personnally tend to prefer using yml format for inventory files though.

File: inventory/sample/hosts.ini [hunk 21]
```diff
@@ -0,0 +1,11 @@
...
+
+[master]
+ubuntu          ansible_host=10.182.31.179    ansible_port=5555   ansible_user=rpi_admin rpi_static_ip=10.0.0.10
```
Comment (rbuisson) [inline inventory/sample/hosts.ini]: This should ideally be dynamically generated in an Ansible task.
But for now, could you provide the default values to be:
```
qemu          ansible_host=10.182.31.179  ansible_user=admin  ansible_port=5555    rpi_static_ip=10.0.0.10
```
or is the hostname of importance and must be `ubuntu`?

File: inventory/sample/hosts.ini [hunk 22]
```diff
@@ -0,0 +1,11 @@
...
+
+[node]
```
Comment (rbuisson) [inline inventory/sample/hosts.ini]: Could you provide one line here as well? To be used as an example.

File: inventory/sample/hosts.ini [hunk 23]
```diff
@@ -0,0 +1,11 @@
...
+
+[k3s_cluster:children]
```
Comment (rbuisson) [inline inventory/sample/hosts.ini]: let's rename this to simply `cluster`

**Review by rbuisson — CHANGES_REQUESTED** [review]
Thanks @chinzou.
I've made few more comments.

Additionally:
- could you provide an ansible.cfg file at the root with:
```
[defaults]
timeout = 30
```
- could you add a task to either inform the user where the iso/img file is, or directly make it accessible on the local machine?
- should we umount/stop anything before copying the iso/img file?

File: roles/qemu/tasks/main.yml [hunk 24]
```diff
@@ -0,0 +1,23 @@
...
+    state: unmounted
+
+- name: Run Qemu VM
+  shell: qemu-system-aarch64 -M virt -m 1024 -smp 4 -cpu cortex-a53 -kernel /opt/vmlinuz -initrd /opt/initrd.img -net tap -net nic -drive file=/opt/os.img,if=none,id=drive0,cache=writeback -device virtio-blk-device,drive=drive0,bootindex=0 -append 'root=/dev/vda2 noresume rw cgroup_enable=cpuset cgroup_memory=1 cgroup_enable=memory' -no-reboot -netdev user,id=net1,hostfwd=tcp::{{ image_ssh_port }}-:22 -device virtio-net-device,netdev=net1 -daemonize -display none
+  become: true
+
+- name: Wait for Qemu VM machine to be available
+  wait_for:
+    port: "{{ image_ssh_port }}"
+    delay: 200
```
Comment (rbuisson) [inline roles/qemu/tasks/main.yml]: Adding:
```
    connect_timeout: 20
    timeout: 400
```
should do.

File: roles/k3s/master/tasks/main.yml (hunk 19 above)
Comment (chinzou) [reply roles/k3s/master/tasks/main.yml]: Thanks for clarifying the naming pattern for Ansible

File: roles/k3s/node/tasks/main.yml [hunk 25]
```diff
@@ -0,0 +1,14 @@
...
+    src: "k3s.service.j2"
+    dest: "{{ systemd_dir }}/k3s-node.service"
+    owner: root
+    group: root
+    mode: 0755
+
+- name: Enable and check K3s service
+  systemd:
+    name: k3s-node
+    enabled: yes
```
Comment (rbuisson) [inline roles/k3s/node/tasks/main.yml:25]: Do we want to make sure it is started as well?

File: roles/os_export/tasks/main.yml [hunk 26]
```diff
@@ -0,0 +1,17 @@
...
+    path: "{{ image_mount_point }}"
+    state: unmounted
+
+- name: Compress the OS image
+  community.general.archive:
+    path: /opt/os.img
+    dest: /opt/os.img.xz
+    format: xz
+
+- name: Fetch the OS image
```
Comment (rbuisson) [inline roles/os_export/tasks/main.yml]: Could you rename to "Retrieve the OS image from the VM"

File: roles/common/tasks/main.yml [hunk 27]
```diff
@@ -1,11 +1,26 @@
 ---
-- debug:
-    msg: "{{ansible_architecture}}"
 
-- name: Update apt cache
+- name: Install network tools
   apt:
-    update-cache: yes
```
Comment (rbuisson) [inline roles/common/tasks/main.yml:7]: Updating the cache is not needed?

File: roles/bootstrap/tasks/main.yml [hunk 28]
```diff
@@ -108,18 +154,39 @@
...
-    path: "{{ image_mount_point }}/dev"
-    src: /dev
-    opts: bind
-    state: mounted
-    fstype: none
+- name: "Create the '{{ cluster_user }}' user in chroot filesystem"
+  shell: chroot {{ image_mount_point }} useradd -u 1041 -m {{ cluster_user }}
+  become: yes
+
+- name: "Create home folder for {{ cluster_user }}"
```
Comment (rbuisson) [inline roles/bootstrap/tasks/main.yml]: Single quote around the cluster_user var so it looks nicer on the log:
"Create home folder for '{{ cluster_user }}'"

File: roles/bootstrap/tasks/main.yml [hunk 29]
```diff
@@ -108,18 +154,39 @@
...
-  ansible.posix.mount:
-    path: "{{ image_mount_point }}/proc"
-    src: /proc
-    opts: bind
-    state: mounted
-    fstype: none
+- name: "Create the '{{ cluster_user }}' user"
+  user:
+    name: "{{ cluster_user }}"
+    comment: Cluster User with root access
```
Comment (rbuisson) [inline roles/bootstrap/tasks/main.yml:160]: "Admin user with root access"

File: roles/bootstrap/tasks/main.yml [hunk 30]
```diff
@@ -1,18 +1,58 @@
...
+- name: Install system packages
   package:
-    name: kpartx
-    state: present
+    name:
+    - kpartx
+    - python3-pip
+    update_cache: yes
+
+- name: Check that os-origin.img exists
```
Comment (rbuisson) [inline roles/bootstrap/tasks/main.yml]: "Check that 'os-origin.img' exists"

File: inventory/sample/hosts.yml [hunk 31]
```diff
@@ -0,0 +1,19 @@
+vm:
```
Comment (rbuisson) [inline inventory/sample/hosts.yml:1]: YAML!! thanks 👍

File: README.md [hunk 32]
```diff
@@ -1,11 +1,38 @@
...
 ```
-ansible-playbook -i hosts.yml  playbook.yml
+
+Retrieve the machine IP
+
+```bash
+multipass info os-builder | grep IPv4
 ```
+
+Edit the [hosts.ini](inventory/sample/hosts.ini) file with the retrieved IP.
```
Comment (rbuisson) [inline README.md]: Edit the [hosts.yml](inventory/sample/hosts.yml) file and set this IP as the `ansible_host` value for all hosts.

File: inventory/sample/hosts.yml [hunk 33]
```diff
@@ -0,0 +1,19 @@
...
+      ansible_user: ubuntu
+master:
+  hosts:
+    ubuntu:
+      ansible_port: 5555
+      ansible_host: 10.182.31.179
+      ansible_user: rpi_admin
+      rpi_static_ip: 10.0.0.10
+node:
+
```
Comment (rbuisson) [inline inventory/sample/hosts.yml]: could you provide 1 node as well?

**Review by rbuisson — CHANGES_REQUESTED** [review]
Thanks. Just minor changes.

File: inventory/sample/hosts.ini (hunk 21 above)
Comment (rbuisson) [reply inventory/sample/hosts.ini]: @chinzou ☝️ ?
> is the hostname of importance and must be ubuntu?

File: roles/common/tasks/main.yml (hunk 27 above)
Comment (chinzou) [reply roles/common/tasks/main.yml:7]: It's been working without it, but maybe it's important for the long term

File: inventory/sample/hosts.yml (hunk 33 above)
Comment (chinzou) [reply inventory/sample/hosts.yml]: For the mean time the playbook can generate only one image at a time.

File: README.md (hunk 32 above)
Comment (chinzou) [reply README.md]: What do you think about setting the IP in the vars file and we can refer to it in the **hosts.yml** file?

Comment (rbuisson) [reply README.md]: Sure. let's do this.

File: inventory/sample/hosts.yml (hunk 33 above)
Comment (rbuisson) [reply inventory/sample/hosts.yml]: As per our discussion, let's try to use a combination of [`serial: 1`](https://stackoverflow.com/a/27315887/11387004)and `delegate_to: vm` to allow processing as many hosts as needed.
Ideally we would like to re-use the Qemu machine between hosts for performance optimization.

Comment (rbuisson) [reply inventory/sample/hosts.yml]: Let this be part of another PR.

File: roles/k3s/node/tasks/main.yml (hunk 25 above)
Comment (rbuisson) [reply roles/k3s/node/tasks/main.yml:25]: Can't be started, since the master is not available.

**Review by rbuisson — APPROVED** [review]
Thanks!
